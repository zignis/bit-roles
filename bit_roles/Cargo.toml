[package]
name = "bit_roles"
version = "0.2.0"
edition = "2021"
authors = ["zignis <velociraptorm77@gmail.com>"]
license = "MIT"
description = "Granular role management based on bit flags"
keywords = ["bit", "enum", "permissions", "roles", "macros"]
documentation = "https://docs.rs/bit_roles"
homepage = "https://github.com/zignis/bit_roles"
repository = "https://github.com/zignis/bit_roles"
readme = "../README.md"

[dependencies]
thiserror = "1.0.58"
bit_roles_macros = { path = "../bit_roles_macros", version = "0.2.0" }