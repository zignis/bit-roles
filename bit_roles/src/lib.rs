//! # Bit Roles
//!
//! [![Latest Version](https://img.shields.io/crates/v/bit_roles.svg)](https://crates.io/crates/bit_roles)
//! [![Rust Documentation](https://docs.rs/bit_roles/badge.svg)](https://docs.rs/bit_roles)
//!
//! This crate enables you to implement granular role and permission management
//! based on bit flags.
//!
//! You can derive the [BitRole] trait for your role enum. It ensures
//! compile-time validation for enum discriminants. Ensure you specify a
//! discriminant for every enum variant; it must be either zero or a power of
//! two. Also, remember to derive the [Copy] and [Clone] traits for your enum.
//!
//! If you need a manager without compile-time checks, it's also exported as
//! [BitRoleUnchecked] trait. This is useful if you want to use raw integer
//! values for roles or have a complex role enum definition. You will need to
//! implement the `Into<usize>` trait for your role enum, along with deriving
//! the [Copy] and [Clone] traits for it.
//!
//! # Examples
//!
//! ```
//! use bit_roles::BitRole;
//!
//! #[derive(Debug)]
//! struct User {
//!     permissions: usize,
//! }
//!
//! #[derive(Debug, BitRole, Copy, Clone)]
//! enum Permission {
//!     None = 0,
//!     SendMessage = 1,
//!     EditMessage = 2,
//! }
//!
//! let mut permissions = Permission::empty();
//! permissions.add_one(Permission::SendMessage);
//!
//! let user = User {
//!     permissions: permissions.get_value(),
//! };
//!
//! // Checking if the user can edit messages.
//! let manager = Permission::from_value(user.permissions);
//! let can_edit_messages = manager.has_one(Permission::EditMessage); // `false`
//! ```

#![forbid(unsafe_code)]
#![warn(future_incompatible, missing_docs)]

mod checked;
mod error;
mod role_value;
mod unchecked;
mod utils;

pub use bit_roles_macros::{
    BitRole,
    BitRoleUnchecked,
};
pub use checked::*;
pub use error::RoleError;
pub use role_value::RoleValue;
pub use unchecked::*;
pub use utils::is_validate_role;

/// The role variant trait. All role enums must implement this trait.
pub trait RoleVariant: Into<usize> + Copy {}
