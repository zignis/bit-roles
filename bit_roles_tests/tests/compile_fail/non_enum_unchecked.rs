use bit_roles::BitRoleUnchecked;

#[derive(Debug, BitRoleUnchecked, Copy, Clone)]
struct Role {}

fn main() {}
