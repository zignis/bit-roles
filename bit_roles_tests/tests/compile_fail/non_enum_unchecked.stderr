error: This macro can only be used with enums.
 --> tests/compile_fail/non_enum_unchecked.rs:3:17
  |
3 | #[derive(Debug, BitRoleUnchecked, Copy, Clone)]
  |                 ^^^^^^^^^^^^^^^^
  |
  = note: this error originates in the derive macro `BitRoleUnchecked` (in Nightly builds, run with -Z macro-backtrace for more info)
